tests/ptvsd/common/test_launcher.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+import os.path
+import pytest
+import subprocess
+import sys
+import time
+import ptvsd.common.launcher
+from ptvsd.common.launcher import parse, WAIT_ON_NORMAL_SWITCH, WAIT_ON_ABNORMAL_SWITCH
+
+launcher_py = os.path.abspath(ptvsd.common.launcher.__file__)
+
+
+@pytest.mark.parametrize("run_as", ["file", "module", "code"])
+@pytest.mark.parametrize("mode", ["normal", "abnormal", "both", ""])
```

Comment:
Nifty trick: you can use `"normal+abnormal"` here instead. Then `split()` it inside the test to get a proper list, and do checks like `if "normal" in mode` etc. That way it only needs two branches; and pytest will still print the names out properly.